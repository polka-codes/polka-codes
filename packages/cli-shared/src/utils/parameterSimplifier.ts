type SimplifiedParams = Record<string, unknown>
type ParameterSimplifier = (params: Record<string, unknown>) => SimplifiedParams
type SimplifierMap = Record<string, ParameterSimplifier>

function replaceInFileSimplifier(params: Record<string, unknown>): SimplifiedParams {
  return { path: params.path }
}

function writeToFileSimplifier(params: Record<string, unknown>): SimplifiedParams {
  return { path: params.path }
}

function readFileSimplifier(params: Record<string, unknown>): SimplifiedParams {
  return { path: params.path, includeIgnored: params.includeIgnored }
}

function listFilesSimplifier(params: Record<string, unknown>): SimplifiedParams {
  const maxCount = params.maxCount
  return {
    path: params.path,
    recursive: params.recursive,
    ...(maxCount !== 2000 && { maxCount }),
  }
}

function searchFilesSimplifier(params: Record<string, unknown>): SimplifiedParams {
  return { ...params }
}

function executeCommandSimplifier(params: Record<string, unknown>): SimplifiedParams {
  return { command: params.command, requiresApproval: params.requiresApproval }
}

function updateMemorySimplifier(params: Record<string, unknown>): SimplifiedParams {
  return { operation: params.operation, topic: params.topic }
}

const SIMPLIFIERS: SimplifierMap = {
  replaceInFile: replaceInFileSimplifier,
  writeToFile: writeToFileSimplifier,
  readFile: readFileSimplifier,
  listFiles: listFilesSimplifier,
  searchFiles: searchFilesSimplifier,
  executeCommand: executeCommandSimplifier,
  updateMemory: updateMemorySimplifier,
}

export function simplifyToolParameters(toolName: string, params: Record<string, unknown> | undefined): Record<string, unknown> {
  if (params === undefined || params === null) {
    return {}
  }

  const simplifier = SIMPLIFIERS[toolName]
  if (simplifier) {
    return simplifier(params)
  }

  return { ...params }
}
