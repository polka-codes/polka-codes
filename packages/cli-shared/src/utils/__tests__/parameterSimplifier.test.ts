import { describe, expect, it } from 'bun:test'
import { simplifyToolParameters } from '../parameterSimplifier'

describe('simplifyToolParameters', () => {
  describe('replaceInFile', () => {
    it('should simplify replaceInFile parameters', () => {
      const result = simplifyToolParameters('replaceInFile', {
        path: 'src/main.ts',
        diff: '<<<<<<< SEARCH\n...very large diff...\n>>>>>>> REPLACE',
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('writeToFile', () => {
    it('should simplify writeToFile parameters', () => {
      const result = simplifyToolParameters('writeToFile', {
        path: 'src/app.ts',
        content: 'export default function App() {\n  return <div>Hello World</div>\n}',
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('readFile', () => {
    it('should simplify readFile parameters', () => {
      const result = simplifyToolParameters('readFile', {
        path: ['src/main.ts', 'src/utils.ts'],
        includeIgnored: false,
      })
      expect(result).toMatchSnapshot()
    })

    it('should handle includeIgnored true', () => {
      const result = simplifyToolParameters('readFile', {
        path: ['src/main.ts'],
        includeIgnored: true,
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('listFiles', () => {
    it('should simplify listFiles parameters with default maxCount', () => {
      const result = simplifyToolParameters('listFiles', {
        path: 'src',
        recursive: true,
        maxCount: 2000,
      })
      expect(result).toMatchSnapshot()
    })

    it('should include maxCount when non-default', () => {
      const result = simplifyToolParameters('listFiles', {
        path: 'src',
        recursive: false,
        maxCount: 100,
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('searchFiles', () => {
    it('should return all searchFiles parameters', () => {
      const result = simplifyToolParameters('searchFiles', {
        path: 'src',
        regex: '^import',
        filePattern: '*.ts',
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('executeCommand', () => {
    it('should simplify executeCommand parameters', () => {
      const result = simplifyToolParameters('executeCommand', {
        command: 'npm run build',
        requiresApproval: false,
      })
      expect(result).toMatchSnapshot()
    })

    it('should handle requiresApproval true', () => {
      const result = simplifyToolParameters('executeCommand', {
        command: 'rm -rf node_modules',
        requiresApproval: true,
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('updateMemory', () => {
    it('should omit content from updateMemory parameters', () => {
      const result = simplifyToolParameters('updateMemory', {
        operation: 'append',
        topic: 'context',
        content: 'This is a very long content that should be omitted from the simplified parameters',
      })
      expect(result).toMatchSnapshot()
    })

    it('should handle replace operation', () => {
      const result = simplifyToolParameters('updateMemory', {
        operation: 'replace',
        topic: 'notes',
        content: 'More content to be omitted',
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('default fallback', () => {
    it('should use default fallback for unknown tools', () => {
      const result = simplifyToolParameters('unknownTool', {
        someParam: 'value',
        anotherParam: 123,
      })
      expect(result).toMatchSnapshot()
    })

    it('should return all parameters for unregistered tool', () => {
      const result = simplifyToolParameters('customTool', {
        config: { option1: true, option2: 'test' },
        data: [1, 2, 3],
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('edge cases', () => {
    it('should handle undefined params', () => {
      const result = simplifyToolParameters('replaceInFile', undefined)
      expect(result).toMatchSnapshot()
    })

    it('should handle null params', () => {
      const result = simplifyToolParameters('writeToFile', null as unknown as undefined)
      expect(result).toMatchSnapshot()
    })

    it('should handle empty params object', () => {
      const result = simplifyToolParameters('readFile', {})
      expect(result).toMatchSnapshot()
    })

    it('should handle params with missing expected fields', () => {
      const result = simplifyToolParameters('replaceInFile', {
        unexpectedField: 'value',
      })
      expect(result).toMatchSnapshot()
    })
  })
})
