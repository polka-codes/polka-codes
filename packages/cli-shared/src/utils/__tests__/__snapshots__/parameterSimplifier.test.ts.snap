// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`simplifyToolParameters replaceInFile should simplify replaceInFile parameters 1`] = `
{
  "path": "src/main.ts",
}
`;

exports[`simplifyToolParameters writeToFile should simplify writeToFile parameters 1`] = `
{
  "path": "src/app.ts",
}
`;

exports[`simplifyToolParameters readFile should simplify readFile parameters 1`] = `
{
  "includeIgnored": false,
  "path": [
    "src/main.ts",
    "src/utils.ts",
  ],
}
`;

exports[`simplifyToolParameters readFile should handle includeIgnored true 1`] = `
{
  "includeIgnored": true,
  "path": [
    "src/main.ts",
  ],
}
`;

exports[`simplifyToolParameters listFiles should simplify listFiles parameters with default maxCount 1`] = `
{
  "path": "src",
  "recursive": true,
}
`;

exports[`simplifyToolParameters listFiles should include maxCount when non-default 1`] = `
{
  "maxCount": 100,
  "path": "src",
  "recursive": false,
}
`;

exports[`simplifyToolParameters searchFiles should return all searchFiles parameters 1`] = `
{
  "filePattern": "*.ts",
  "path": "src",
  "regex": "^import",
}
`;

exports[`simplifyToolParameters executeCommand should simplify executeCommand parameters 1`] = `
{
  "command": "npm run build",
  "requiresApproval": false,
}
`;

exports[`simplifyToolParameters executeCommand should handle requiresApproval true 1`] = `
{
  "command": "rm -rf node_modules",
  "requiresApproval": true,
}
`;

exports[`simplifyToolParameters updateMemory should omit content from updateMemory parameters 1`] = `
{
  "operation": "append",
  "topic": "context",
}
`;

exports[`simplifyToolParameters updateMemory should handle replace operation 1`] = `
{
  "operation": "replace",
  "topic": "notes",
}
`;

exports[`simplifyToolParameters default fallback should use default fallback for unknown tools 1`] = `
{
  "anotherParam": 123,
  "someParam": "value",
}
`;

exports[`simplifyToolParameters default fallback should return all parameters for unregistered tool 1`] = `
{
  "config": {
    "option1": true,
    "option2": "test",
  },
  "data": [
    1,
    2,
    3,
  ],
}
`;

exports[`simplifyToolParameters edge cases should handle undefined params 1`] = `{}`;

exports[`simplifyToolParameters edge cases should handle null params 1`] = `{}`;

exports[`simplifyToolParameters edge cases should handle empty params object 1`] = `
{
  "includeIgnored": undefined,
  "path": undefined,
}
`;

exports[`simplifyToolParameters edge cases should handle params with missing expected fields 1`] = `
{
  "path": undefined,
}
`;
